pkg/reconciler/ingress/status.go

```diff
@@ -277,63 +370,65 @@ func (m *StatusProber) processWorkItem() bool {
-		namespace, name, err := cache.SplitMetaNamespaceKey(name)
-		if err != nil {
-			return nil, fmt.Errorf("failed to parse Gateway name %q: %v", name, err)
-		}
-		if namespace == "" {
-			namespace = vs.Namespace
-		}
-		gateway, err := m.gatewayLister.Gateways(namespace).Get(name)
-		if err != nil {
-			return nil, fmt.Errorf("failed to get Gateway %s/%s: %v", namespace, name, err)
+		// This is the last pod being successfully probed, the Ingress is ready
+		if atomic.AddInt32(&ingressState.pendingCount, -1) == 0 {
+			m.readyCallback(ingressState.ia)
 		}
+	}
+}
 
-		// Skip gateways that cannot be probed
-		if !resources.CanProbeGateway(gateway) {
-			continue
-		}
+func (m *StatusProber) getGateway(name string) (*v1alpha3.Gateway, error) {
+	namespace, name, err := cache.SplitMetaNamespaceKey(name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse Gateway name %q: %v", name, err)
+	}
+	if namespace == "" {
+		return nil, fmt.Errorf("unexpected unqualified Gateway name %q", name)
+	}
+	return m.gatewayLister.Gateways(namespace).Get(name)
+}
 
-		// List matching Pods
-		selector := labels.NewSelector()
-		for key, value := range gateway.Spec.Selector {
-			requirement, err := labels.NewRequirement(key, selection.Equals, []string{value})
-			if err != nil {
-				return nil, fmt.Errorf("failed to create 'Equals' requirement from %q=%q: %v", key, value, err)
-			}
-			selector = selector.Add(*requirement)
-		}
-		pods, err := m.podLister.List(selector)
+func (m *StatusProber) listGatewayPods(gateway *v1alpha3.Gateway) ([]*corev1.Pod, error) {
+	selector := labels.NewSelector()
+	for key, value := range gateway.Spec.Selector {
+		requirement, err := labels.NewRequirement(key, selection.Equals, []string{value})
 		if err != nil {
-			return nil, fmt.Errorf("failed to list Pods: %v", err)
+			return nil, fmt.Errorf("failed to create 'Equals' requirement from %q=%q: %v", key, value, err)
 		}
+		selector = selector.Add(*requirement)
+	}
+	pods, err := m.podLister.List(selector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list Pods: %v", err)
+	}
+	return pods, nil
+}
 
-		// Filter out the Pods without an assigned IP address
-		for _, pod := range pods {
-			if pod.Status.PodIP != "" {
-				podIPs = append(podIPs, pod.Status.PodIP)
+// listURLTemplates returns a list of string templates (contains one "%s") to be used to generate probe URLs
+// from pod IPs. For example, if the Gateway exposes an HTTPS endpoint on port 8080, this will return "https://%s:8080/".
+func (m *StatusProber) listURLTemplates(gateway *v1alpha3.Gateway) []string {
+	templates := make([]string, 0, len(gateway.Spec.Servers))
+	for _, server := range gateway.Spec.Servers {
+		switch server.Port.Protocol {
+		case v1alpha3.ProtocolHTTP, v1alpha3.ProtocolHTTP2:
+			if server.TLS == nil || !server.TLS.HTTPSRedirect {
+				templates = append(templates, "http://%s:"+strconv.Itoa(server.Port.Number))
```

Comment:
Are you sure that this port number actually points to the port of the pod and not to the port of the service of the pods?

In my cluster, the gateway looks like this:

```yaml
Name:         cluster-local-gateway
Namespace:    knative-serving
...
Spec:
  Selector:
    Istio:  cluster-local-gateway
  Servers:
    Hosts:
      *
    Port:
      Name:      http
      Number:    80
      Protocol:  HTTP
```

And the actual endpoints look like this

```yaml
apiVersion: v1
kind: Endpoints
...
subsets:
- addresses:
  - ip: 10.130.2.5
...
  ports:
  - name: https
    port: 8443
    protocol: TCP
  - name: tls
    port: 15443
    protocol: TCP
  - name: status-port
    port: 15020
    protocol: TCP
  - name: http2
    port: 8080
    protocol: TCP
```

So I think we actually need to resolve ports via the Service or rather go directly via Endpoints which gives us proper ports + endpoints list. See #5531 for a potential first step towards that.

Reply:
You are correct.